rfcs/0048-on-conflict-clause-in-create-table-statement.md

```diff
@@ -0,0 +1,121 @@
+---
+feature: ON CONFLICT Clause in Create Table Statement
+authors:
+  - "st1page"
+start_date: "2023/2/1"
+---
+
+# ON CONFLICT Clause in Create Table Statement
+
+## Summary
+
+allow user to declare the conflict behavior when the newly inserted row break the unique constraint of primary key.
+
+## Motivation
+
+Currently, for a table with primary key, the later operations will always overwrite the previous ones when the new operation [0017-DML.md](0017-DML.md)
+But users' requirement are various here.
+I will give some different behavior here. Notice that here the conflict behavior is not just happens when an insert DML statement happening but also in a table with external connector.
+
+- **Overwrite**
+  The most important needs of using the Overwrite behavior is to express "Upsert" operation. Some connector can not give the old value of the update operation and we need to overwrite the old value with the same pk in storage.
+
+- **Ignore**
+  The Ignore behavior make user can define a "append-only table with primary key" which maintain the primary constrain and also provide a append-only changes to downstream with good performance. This is good to express a table with an At-least-once Delivery connector where a same record can be replayed multiple times to break the primary key constrain.
+  Notice: the specific grammar for this has not designed(e.g. should we expose append-only table to user?) but the situation is the same.
+
+- **Partial Update**
+  The ON CONFLICT make users can do partial update on the table.
+
+  ```SQL
+    CREATE table t(k int primary key, int a, int b) 
+    ON CONFLICT DO UPDATE SET (a,b) = (
+      COALESCE(EXCLUDED.a, a),
+      COALESCE(EXCLUDED.b, b));
+  ```
+  
+  This feature will be useful when user use wide-table schema data model.
```

Comment:
As I understand it, this is similar to Apache Doris' pre-aggregate data model, where the data of the same PK can be automatically merged according to some custom rules. However, as one of its selling points, Doris has a specially designed storage to do this pre-aggregation efficiently, but we don't have it.

Apparently, our storage is optimized for non-conflict use cases, which means that most rows will not have a conflict on PK, and the PK check in MaterializeExecutor is just a precaution to avoid a crash. On the contrary, this RFC assumes that incoming rows will have many conflicts by design, and hopes to resolve them in MaterializeExecutor. I don't think it's a good idea to encourage users to do this, and it's not worth optimizing for.

Also, I am not very confident that checking in MaterializeExecutor is a good idea. Maybe it will become a painpoint at some point and maybe users will ask us to remove it. I don't really recommend relying on it much.

Reply:
Anyway, we can just treat the check in the MaterializeExecutor as a pluggable feature. Now it is implemented as a flag in MaterializeExecutor, so if users ask us to remove it we can remove it easily.

Reply:
> As I understand it, this is similar to Apache Doris' pre-aggregate data model, where the data of the same PK can be automatically merged according to some custom rules. However, as one of its selling points, Doris has a specially designed storage to do this pre-aggregation efficiently, but we don't have it.

https://doris.apache.org/docs/data-table/data-model/#aggregate-model just supports some limited behaviors. 
```
AggregationTypecurrently has the following four ways of aggregation:
SUM: Sum, multi-line Value accumulation.
REPLACE: Instead, Values in the next batch of data will replace Values in rows previously imported.
MAX: Keep the maximum.
MIN: Keep the minimum.

``` 
In fact, if we just want to support this behavior, RisingWave has already been able to express it with a `Create Materialized view` with an aggregate query. 
The `on conflicts` can take much more expressiveness to users.